x/auth/ante.go

```diff
@@ -35,7 +35,11 @@ func NewAnteHandler(am AccountMapper, fck FeeCollectionKeeper) sdk.AnteHandler {
 		}
 
 		// set the gas meter
-		newCtx = ctx.WithGasMeter(sdk.NewGasMeter(stdTx.Fee.Gas))
+		if stdTx.Fee.Gas == 0 {
+			newCtx = ctx.WithGasMeter(sdk.NewInfiniteGasMeter())
```

Comment:
Why is this here? Using an infinite gas meter here seems odd to me.

Reply:
When running in simulation mode, we want no gas capping  whatsoever - CC'ing @cwgoes

Reply:
We don't need to cap gas in a simulation, we just need to measure it - is there a reason you think we would need to cap it @ValarDragon?

Reply:
I don't like switching on gas equals 0. This feels like it could cause more bugs / increase the attack surface. I think we should find another way of communicating that were in simulation mode.

Reply:
Could it not be set in the context somewhere upstream?

Reply:
I guess one could add a `CLIContext` to the `NewAnteHandler()`'s arguments - or even a `GasMeter`. That would be set a Baseapp initialization time, which could work for CLI commands yet cause issues to the REST endpoints.

Reply:
What happens if I make a tx with 0 gas, do I suddenly get infinite gas? Is there some sequence of conditionals that can let a 0 gas tx get through? etc. 

I think having a simulate mode being recognized from something separate from gas is safer, and results in more cleared / maintanable code. There are two routes here, either use the existing NewAnteHandler for both the simulate and normal case, or use a seperate ante handler creation function in simulation mode. I'm not opposed to the latter, especially because simulations shouldn't be verifying signatures or depending on the users password.

Reply:
What if we use a default gas limit (some sort of sane default, e.g. `200000` - which is current one) and switch back to use a capped `GasMeter`?

Reply:
Re:
> simulations shouldn't be verifying signatures

That too is tricky. Simulations go through the query interface, which requires messages to be signed.

Reply:
That still doesn't really alleviate my concern. (That does limit the situation though)

I do think switching on 0 gas is less safe, as we now have to make sure that a tx on the network with 0 gas is stopped before getting to this line. Its prefferrable to just have a separate method to flag simulations, and have the normal gas logic prevent a 0 gas tx imo.

Reply:
Do you reckon the following could mitigate the risk?

```diff
diff --git a/client/utils/utils.go b/client/utils/utils.go
index fb5d6198..49258b86 100644
--- a/client/utils/utils.go
+++ b/client/utils/utils.go
@@ -1,6 +1,7 @@
 package utils
 
 import (
+       "errors"
        "fmt"
        "os"
 
@@ -108,6 +109,10 @@ func CalculateGas(queryFunc func(string, common.HexBytes) ([]byte, error), cdc *
        if err != nil {
                return
        }
+       if estimate <= 0 {
+               err = errors.New("simulation returned gas=0")
+               return
+       }
        adjusted = adjustGasEstimate(estimate, adjustment)
        fmt.Fprintf(os.Stderr, "gas: [estimated = %v] [adjusted = %v]\n", estimate, adjusted)
        return

```